Prevent mixing deprecated check-cfg syntax with new one

Urgau opened the pull request:
This PR puts an hard error when trying to mix the deprecated check-cfg syntax with new one, doing that required updating the `libc` crate crate to version 0.2.150 (this also means making use of the new syntax when building stage 2 and above as `libc` isn't granular enough).

This was previously attempted in https://github.com/rust-lang/rust/pull/117025 but had to be striped out due to the `libc` crate not being compliant.

While preventing the mixing isn't necessary per se, it will help make sure that we are going to be ready when removing the deprecated syntax.

r? @petrochenkov

Urgau committed:
check-cfg: prevent mixing the deprecated syntax with new one

Urgau committed:
libc: bump dependency to 0.2.150

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6761962248/job/18377372651) [(plain)](https://github.com/rust-lang/rust/commit/0383ca39177a51eaef1f90907d3f62b1dd928fdf/checks/18377372651/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=Urgau
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_d7400db7-f21d-45fb-855a-2180cb43c949
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=check-cfg-prevent-mixing
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_d7400db7-f21d-45fb-855a-2180cb43c949
GITHUB_REF=refs/pull/117612/merge
GITHUB_REF_NAME=117612/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=1d0f6525ede749bde59ef3391f21946e38cfeb50
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_d7400db7-f21d-45fb-855a-2180cb43c949
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_d7400db7-f21d-45fb-855a-2180cb43c949
GITHUB_TRIGGERING_ACTOR=Urgau
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/117612/merge
GITHUB_WORKFLOW_SHA=1d0f6525ede749bde59ef3391f21946e38cfeb50
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
   Compiling bitflags v2.4.0
   Compiling linux-raw-sys v0.4.10
   Compiling parking_lot_core v0.8.6
   Compiling instant v0.1.12
error: invalid `--check-cfg` argument: `cfg(emscripten_new_stat_abi)` (cannot mix `cfg(...)` with deprecated syntax `names(...)` and `values(...)`)
error: could not compile `libc` (lib) due to previous error
warning: build failed, waiting for other jobs to finish...
Build completed unsuccessfully in 0:04:37
cat: /tmp/toolstate/toolstates.json: No such file or directory
```

</details>

Urgau committed:
bootstrap: use new check-cfg syntax when building stage 2 and above

This is required by the libc crate since it uses the new syntax as soon
as it detects rustc version 1.75 or above, which is the case when
building with stage 1 and above.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6762232790/job/18377954653) [(plain)](https://github.com/rust-lang/rust/commit/12a3873b0e550dbe31e3426787233911b5565a4c/checks/18377954653/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=Urgau
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_84f04e6a-4de2-4b92-9ed3-4ff9a5d91309
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=check-cfg-prevent-mixing
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_84f04e6a-4de2-4b92-9ed3-4ff9a5d91309
GITHUB_REF=refs/pull/117612/merge
GITHUB_REF_NAME=117612/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=a53184d0cd31b38d615d4e6eb211235e49727e55
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_84f04e6a-4de2-4b92-9ed3-4ff9a5d91309
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_84f04e6a-4de2-4b92-9ed3-4ff9a5d91309
GITHUB_TRIGGERING_ACTOR=Urgau
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/117612/merge
GITHUB_WORKFLOW_SHA=a53184d0cd31b38d615d4e6eb211235e49727e55
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
##[group]Building stage1 library artifacts (x86_64-unknown-linux-gnu)
error: failed to run `rustc` to learn about target-specific information

Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustc - --crate-name ___ --print=file-names --crate-type bin --crate-type rlib --crate-type dylib --crate-type cdylib --crate-type staticlib --crate-type proc-macro --print=sysroot --print=split-debuginfo --print=crate-name --print=cfg` (exit status: 1)
  --- stderr
  error: invalid `--check-cfg` argument: `cfg(bootstrap)` (expected `names(name1, name2, ... nameN)` or `values(name, "value1", "value2", ... "valueN")`)
Build completed unsuccessfully in 0:03:17
  local time: Sun Nov  5 15:41:20 UTC 2023
  network time: Sun, 05 Nov 2023 15:41:20 GMT
##[error]Process completed with exit code 1.
```

</details>

Urgau commented:
Hum. This may not be possible yet, I forgot that we still use beta Cargo, even for stage 2 and above.

@rustbot author
@rustbot blocked

Urgau commented:
I don't think this is is gonna work out before the next Cargo bump, at which point we will remove the old syntax, rendering this PR useless.

Closing for now.
